hlapipeline/utils/astrometric_utils.py

```diff
@@ -314,63 +355,299 @@ def extract_sources(img, fwhm=2.0, threshold=None, source_box=7,
 
     Returns
     --------
-    index : int
-        Index of source from catalog list for source furthest from all others
+    customwcs : object
+        HSTWCS object which contains the undistorted WCS representing the entire
+        field-of-view for the input image
 
-    nearest_dist : float
-        Distance from isolated source to it's nearest neighbor in the catalog
+    Syntax
+    -------
+    This function can be used with the following syntax to apply a shift/rot/scale
+    change to the same image:
 
-    .. note ::
-        This function assumes that the catalog has already been limited to
-        only sources which overlap the actual image area.
+    >>> import buildref
+    >>> from drizzlepac import updatehdr
+    >>> filename = "jce501erq_flc.fits"
+    >>> wcslin = buildref.build_self_reference(filename)
+    >>> updatehdr.updatewcs_with_shift(filename,wcslin,xsh=49.5694, ysh=19.2203, rot = 359.998, scale = 0.9999964)
 
     """
-    if columns:
-        cat1 = np.column_stack((catalog[columns[0]], catalog[columns[1]]))
+    if 'sipwcs' in filename:
+        sciname = 'sipwcs'
     else:
-        cat1 = catalog
-    cat_dist = distance_matrix(cat1, cat1)
-    y_max = cat_dist.sum(axis=1)
-    iso_indx = y_max.argsort()[-1]
+        sciname = 'sci'
 
-    # get index to next nearest neighbor to this isolated source
-    next_nearest = cat_dist[iso_indx].argsort()[1]
-    # now remember the distance to that source from the isolated source
-    iso_dist = cat_dist[iso_indx][next_nearest]
+    wcslin = build_reference_wcs([filename], sciname=sciname)
+
+    if clean_wcs:
+        wcsbase = wcslin.wcs
+        customwcs = build_hstwcs(wcsbase.crval[0],wcsbase.crval[1],wcsbase.crpix[0],wcsbase.crpix[1],wcslin._naxis1,wcslin._naxis2,wcslin.pscale,wcslin.orientat)
+    else:
+        customwcs = wcslin
+    return customwcs
+
+def read_hlet_wcs(filename, ext):
+    """Insure HSTWCS includes all attributes of a full image WCS.
+
+    For headerlets, the WCS does not contain information about the size of the
+    image, as the image array is not present in the headerlet.
+    """
+    hstwcs = wcsutil.HSTWCS(filename, ext=ext)
+    if hstwcs.naxis1 is None:
+        hstwcs.naxis1 = int(hstwcs.wcs.crpix[0]*2.) # Assume crpix is center of chip
+        hstwcs.naxis2 = int(hstwcs.wcs.crpix[1]*2.)
 
-    return iso_indx, iso_dist
+    return hstwcs
 
+def build_hstwcs(crval1, crval2, crpix1, crpix2, naxis1, naxis2, pscale, orientat):
+    """ Create an HSTWCS object for a default instrument without distortion
+        based on user provided parameter values.
+
+        .. note :: COPIED from drizzlepac.wcs_functions
+    """
+    wcsout = wcsutil.HSTWCS()
+    wcsout.wcs.crval = np.array([crval1,crval2])
+    wcsout.wcs.crpix = np.array([crpix1,crpix2])
+    wcsout.naxis1 = naxis1
+    wcsout.naxis2 = naxis2
+    wcsout.wcs.cd = buildRotMatrix(orientat)*[-1,1]*pscale/3600.0
+    # Synchronize updates with PyWCS/WCSLIB objects
```

Comment:
I am just making a comment here as I am reminded (nothing for you to change), but it seems to me there should be some pre-defined constants (e.g., RADTODEG, etc.) which live in some class and can be used by this entire project (or other projects) so hard-coded factors do not have to be used, and the same values do not need to be recomputed over and over again.  I see your methods in astrometric_utils.py which are most helpful, but they could be constants instead and then used as descriptive multiplicative factors.